doc/counters_enabling_redesign.md

```diff
@@ -0,0 +1,66 @@
+# Counters Enabling Redesign
+
+## Table of Content
+
+* [Revision](#revision)
+* [Overview](#overview)
+* [Requirements](#requirements)
+* [High-Level Design](#high-level-design)
+
+
+### Revision
+
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 |  12.15.2021 | Noa Or             | Initial version                   |
+
+
+## Overview
+
+In the current design of counter enabling, when enable_counters.py script is triggered, if the switch uptime is less then 5 minutes, it sleeps for 3 minutes. This is done in order to make sure the system is up and synchronized.
+
+The purpose of counters redesign is to change the 'sleep' mechanisem to be event driven and to have the exact time when the switch is ready to enable counters.
+In addition, enable_counters.py script will be removed, and its logic will now be in flexCountersOrch.
+
+## Requirements
+
+Remove enable_counters.py script and merge the logic into FlexCountersOrch.
+FlexCountersOrch will wait for system to be up using events from APP DB, then enable the counters using the logic written in enable_counters.py.
+
+## High-Level Design
+
+FlexCountersOrch daemon will be refactored to do the following:
+
+- Query admin_status for each port from config_db - PORTCHANNEL & PORT tables.
+
+- Create an intenral data structure that will contain:
+
+    { port: [admin_status, oper_status] }
+    for all ports received from config_db. oper_status will be initialized as an empty string.
+
+- Start listen to events from APP DB, PORT_TABLE, LAG_TABLE.
+
+    For each event arriving:
+    - If Port table doesn't have "PortConfigDone" key, delete the event and continue to the next one.
+    - If the current oper_status of the port equals to the expected, remove the port from the list.
+    - Otherwise, continue.
+
+- In order to handle changes of changing port's state, or using DPC capabilities,
+    listen to events from config_db, PORT & PORTCHANNEL tables.
+    for each event arriving:
+
+    - If a port state has been changed, change it in the internal data structure.
+    - If a port/portchannel was added, add it to the internal data structure with the proper admin status.
+    - If a port/portchannel was deleted, remove it from the internal data structure.
+    - If a delete operation was performed to one of the ports inside the internal data structure - remove the port.
+
+Eventually, when the list is empty, enable the counters.
+
+- The daemon will also wait for a timer in order to be able to enable counters even if one of the ports is not stable.
+    
+    If after 3 minutes (180 seconds), the counters were not enabled yet, enable counters.
+
+    The daemon will create a SelectableTimer and wait for the it to poke.
+
+    When doTask(SelectableTimer*) function will be triggered, it will check if the counters are already enabled.
+    If not, enable them.
```

Comment:
> Chances of race-condition here? What happens if counters enabling is in progress and the `SelectableTimer` is also triggered?

If counters enabling is in progress, it means the daemon first changed a boolean to True. so when the timer will expire, it will check the boolean and will not enable the counters as well.

> Also, after the enabling of counters, what happens to the new data structure defined in `FlexCountersOrch`? There can be two cases to be covered here - event-driven mechanism (new design) and timeout (old design).

Not sure I understand the concern here. we are using the event driven mechanism, but also don't want to be blocked by one flipping port for example, so we have timer to cover this case.